federatedscope/attack/auxiliary/backdoor_utils.py

```diff
@@ -0,0 +1,450 @@
+    img[height - 1][width - 1] = 255
+    img[height - 1][width - 2] = 0
+    img[height - 1][width - 3] = 255
+
+    img[height - 2][width - 1] = 0
+    img[height - 2][width - 2] = 255
+    img[height - 2][width - 3] = 0
+
+    img[height - 3][width - 1] = 255
+    img[height - 3][width - 2] = 0
+    img[height - 3][width - 3] = 0
+
+    # left top
+    img[1][1] = 255
+    img[1][2] = 0
+    img[1][3] = 255
+
+    img[2][1] = 0
+    img[2][2] = 255
+    img[2][3] = 0
+
+    img[3][1] = 255
+    img[3][2] = 0
+    img[3][3] = 0
+
+    # right top
+    img[height - 1][1] = 255
+    img[height - 1][2] = 0
+    img[height - 1][3] = 255
+
+    img[height - 2][1] = 0
+    img[height - 2][2] = 255
+    img[height - 2][3] = 0
+
+    img[height - 3][1] = 255
+    img[height - 3][2] = 0
+    img[height - 3][3] = 0
+
+    # left bottom
+    img[1][width - 1] = 255
+    img[2][width - 1] = 0
+    img[3][width - 1] = 255
+
+    img[1][width - 2] = 0
+    img[2][width - 2] = 255
+    img[3][width - 2] = 0
+
+    img[1][width - 3] = 255
+    img[2][width - 3] = 0
+    img[3][width - 3] = 0
+
+    return img
+
+
+def _fourCorner_w_Trigger( img, height, width, distance, trig_h, trig_w):
+    # right bottom
+    img[height - 1][width - 1] = 255
+    img[height - 1][width - 2] = 255
+    img[height - 1][width - 3] = 255
+
+    img[height - 2][width - 1] = 255
+    img[height - 2][width - 2] = 255
+    img[height - 2][width - 3] = 255
+
+    img[height - 3][width - 1] = 255
+    img[height - 3][width - 2] = 255
+    img[height - 3][width - 3] = 255
+
+    # left top
+    img[1][1] = 255
+    img[1][2] = 255
+    img[1][3] = 255
+
+    img[2][1] = 255
+    img[2][2] = 255
+    img[2][3] = 255
+
+    img[3][1] = 255
+    img[3][2] = 255
+    img[3][3] = 255
+
+    # right top
+    img[height - 1][1] = 255
+    img[height - 1][2] = 255
+    img[height - 1][3] = 255
+
+    img[height - 2][1] = 255
+    img[height - 2][2] = 255
+    img[height - 2][3] = 255
+
+    img[height - 3][1] = 255
+    img[height - 3][2] = 255
+    img[height - 3][3] = 255
+
+    # left bottom
+    img[1][width - 1] = 255
+    img[2][width - 1] = 255
+    img[3][width - 1] = 255
+
+    img[1][width - 2] = 255
+    img[2][width - 2] = 255
+    img[3][width - 2] = 255
+
+    img[1][height - 3] = 255
+    img[2][height - 3] = 255
+    img[3][height - 3] = 255
+
+    return img
+
+
+def _randomPixelTrigger( img, height, width, distance, trig_h, trig_w):
+    alpha = 0.2
+    mask = np.random.randint(low=0, high=256, size=(height, width), dtype=np.uint8)
+    blend_img = (1 - alpha) * img + alpha * mask.reshape((height, width, 1))
+    blend_img = np.clip(blend_img.astype('uint8'), 0, 255)
+
+    return blend_img
+
+
+def _signalTrigger( img, height, width, distance, trig_h, trig_w):
+    #  vertical stripe pattern different from sig
+    alpha = 0.2
+    # load signal mask
+    signal_mask = np.load('/mnt/zeyuqin/FederatedScope/federatedscope/attack/triggers/signal_cifar10_mask.npy')
```

Comment:
Please not hard code the file path. Please check all the file paths.

Reply:
fixed it